Impl for Parameter Dict

arktrail opened the pull request:
Try to implement the `nn::ParameterDict` for the C++ front-end from #25883.

@yf225  please take a look and tell me if the PR has problems. I took the #36904 and #28652 as reference. I am hoping for any suggestions.

arktrail committed:
Impl for Parameter Dict

arktrail committed:
Impl for Parameter Dict

arktrail committed:
Impl for Parameter Dict

arktrail committed:
Impl for Parameter Dict

arktrail committed:
Impl for Parameter Dict

arktrail committed:
Impl for Parameter Dict

arktrail committed:
Impl for Parameter Dict

yyn19951228 commented:
When implementing the ParameterDict, I refer to ModuleDict from #28652, in which the ModuleDict derived from `Cloneable`. It seems that the `Cloneable` is just for `Module`, so I guess when it comes to the Parameter part (both ParameterDict and ParameterList), I don't need to derive the `Cloneable`?

It is my first time to commit to PyTorch, so please correct me if my understanding is totally wrong, thank you guys so much!

glaringlee commented:
@yyn19951228 
Thanks for working on this. I think there are some misunderstanding here. Let me explain:
Overall, we treat the parameterDict/List like a Module.

1. ParameterDict should be inherited from Module class and clone-able (What you've done is right). The reason is that we want parameterDict be visible by all the Module functions, and clone-able. So parameterDict can be used as a normal dict, it can also be used as a Module if necessary. We don't have any use case though, but we should make it consistent with python impl.

2. You don't need to create a new OrderedDict in your ParameterDict class, instead, you should reuse the parameter OrderedDict within Module class, so all other functions in Module functions can see the parameterDict, and clone() works in that case.

Take a look at the python implementation here:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/container.py#L485
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py (as reference)
We have module c++ impl here and it support most of the functionalities in module.py
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/module.h
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/src/nn/module.cpp

What you need to do, is to follow the parameterDict python impl translate it into C++. I think most of the function will be exactly the. same as python implementation. The update() api, you might need to overload it into different version. (Iterator, dict, etc). Let me know if you need any help.

glaringlee reviewed (changes requested):
See my comment above.

yyn19951228 commented:
@glaringlee Many thanks for reviewing my diff! I do have some misunderstanding about the implementations, let me rewrite these parts.

arktrail committed:
Impl for ParameterDict V2

yyn19951228 commented:
@glaringlee I have updated my impl for parameterDict, please check if it meet the necessity. Also, I get lost for two places:
1. For the initialization function, I would like to override by taking the `std::initializer_list` like this
```  
explicit ParameterDictImpl(
      std::initializer_list<std::pair<std::string, Tensor>> initializer_list) {
    parameters_.reserve(initializer_list.size());
    for (const auto& item : initializer_list) {
      insert(std::move(item.first), std::move(item.second));
    }
  }
```
But when I want to init by using
```
ParameterDict dict(
    {
      {"a", torch::tensor({1.0})}
    }
  );
```
it shows that 
```
no matching constructor for initialization of 'torch::nn::ParameterDict'ccls(2)
pimpl.h(53, 18): candidate inherited constructor not viable: cannot convert initializer list argument to 'std::nullptr_t' (aka 'nullptr_t')
(162, 1): constructor from base class 'ModuleHolder<torch::nn::ParameterDictImpl>' inherited here
pimpl.h(70, 18): candidate inherited constructor not viable: cannot convert initializer list argument to 'std::shared_ptr<ParameterDictImpl>'
.....
no instance of constructor "torch::nn::ParameterDict::ParameterDict" matches the argument list -- argument types are: ({...})
```
May I ask what might be the proper way to implement this?

2. I use the `container` as the arguments for `update()`, may I ask does `iterator` also needed in such case?

Thank you very much! And sorry for these questions, they seem dummy maybe.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+        "No Parameter with name `",
+        key,
+        "` is registered");
+    torch::Tensor v = parameters_[key].clone();
```
You don't need to clone(), we just need to return the same tensor, if you clone() the underlying storage will be copied. See below, it is not copied.
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/container.py#L557

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+  /// Return the keys in the dict
+  ::std::vector<std::string> keys() const {
+    std::vector<std::string> keys;
+    keys.reserve(parameters_.size());
```
I think you can call this directly. 
https://github.com/pytorch/pytorch/blob/a203dc2e6d74c8abb63145873b757769d1383273/torch/csrc/api/include/torch/ordered_dict.h#L163

yyn19951228 replied:
Oh, I missed that.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+  /// Return the Values in the dict
+  ::std::vector<torch::Tensor> values() const {
+    std::vector<torch::Tensor> values;
+    values.reserve(parameters_.size());
```
Same as above
https://github.com/pytorch/pytorch/blob/a203dc2e6d74c8abb63145873b757769d1383273/torch/csrc/api/include/torch/ordered_dict.h#L167

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+
+  /// Returns an `OrderedDict` with the parameters of this `Module` along with
+  /// their keys, and if `recurse` is true also recursively of every submodule.
+  OrderedDict<std::string, Tensor> named_parameters(bool recurse = true) const {
```
Please don't expose this function, since named_parameters() actually will recursively retrieve submodules' parameters as well when recurse == true. This class only deal with parameter_, so we don't need this function be exposed.. 
And since this class is inherited from Module, we actually have this function automatically.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+      if (contains(item.key())) {
+        pop(item.key());
+      }
+      insert(std::move(item.key()), std::move(item.value()));
```
Same thing here as in the constructor. I think you should insert into parameter_ directly to avoid setting require_grad to true as default.

You don't need to check contains, I think you can just do parameters_[item.key()] = item.value(). This will do existence check and write (or overwrite) the new data.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+    return parameters_.contains(key);
+  }
+
+  Tensor& get(std::string key) {
```
const std::string& key

glaringlee commented on test/cpp/api/parameterdict.cpp:
```diff
@@ -0,0 +1,114 @@
+#include <gtest/gtest.h>
```
Please modify the test if necessary after you finish modifying the parameterdict implementation.

glaringlee reviewed (changes requested):
@yyn19951228 Thank you so much for working on this. I've made some comments. Please take a look. I pasted some links in the comments, so you know why the code should work like that. Let me know if you need any help, I got some env issue today, can not build libtorch for long time, sorry for replying late.

glaringlee commented:
Now answer your previous two questions.
1. You can not use initialize_list at this point. We actually use a ModuleHolder class to hold your implementation. So the initialize_list passed to ModuleHolder first and then it will be forwarded to the ParameterDict impl class. but unfortunately, we have no correct template to forward a single initialize_list at this point. see here. https://github.com/pytorch/pytorch/blob/63c3b89c1cf3fd646d0f8c4bb4aa4ad8cae5d4aa/torch/csrc/api/include/torch/nn/pimpl.h#L57

2. We don't need to support Container in iterator since the iterator is for ParameterDict itself, the only type is OrderedList (parameter_).

yyn19951228 commented on torch/csrc/api/include/torch/nn/module.h:
```diff
@@ -547,6 +547,8 @@ class TORCH_API Module : public std::enable_shared_from_this<Module> {
   template <typename ModuleType, typename... ArgumentTypes>
   friend struct AnyModuleHolder;
 
+  friend class ParameterDictImpl;
```
That makes more sense to me! I am avoiding changing the `module.h`, so instead of changing the access of `parameters_`, I add the friend class instead.

arktrail committed:
Merge branch 'master' of https://github.com/yyn19951228/pytorch

ParameterDict V3

yyn19951228 commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,157 @@
+      torch::OrderedDict<std::string, torch::Tensor> params) {
+    parameters_.reserve(params.size());
+    for (const auto& item : params) {
+      insert(std::move(item.key()), std::move(item.value()));
```
Compiler complains 
```
torch::OrderedDict<std::__1::string, at::Tensor> torch::nn::Module::parameters_
The registered parameters of this `Module`.
Inorder to access parameters_ in ParameterDict and ParameterList

type 'OrderedDict<std::string, at::Tensor>' (aka 'OrderedDict<basic_string<char, char_traits<char>, allocator<char> >, at::Tensor>') does not provide a call operatorccls(2)
call of an object of a class type without appropriate operator() or conversion functions to pointer-to-function type```
if I use `parameters_(params)` here. 
So can I use `parameters_ = params` here? I think `OrderedDict` implements the copy constructor.
https://github.com/pytorch/pytorch/blob/f4365cf5ba8ee4ba319385cbd2ff7b9a825a75d0/torch/csrc/api/include/torch/ordered_dict.h#L40

glaringlee commented:
@yyn19951228 stuck with another code review today, will review yours tomorrow if I can not make it today.

yyn19951228 commented:
@glaringlee Doesn't matter.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,135 @@
+
+  /// Remove key from the ParameterDict, and return its value
+  Tensor pop(const std::string& key) {
+    TORCH_CHECK(
```
We don't need this TORCH_CHECK, since the ordered_dict.h already did this. parameters_[key] will throw if key is not in the parameters_
https://github.com/pytorch/pytorch/blob/a54416d208cd024779455c7b44e7b86f51a3da45/torch/csrc/api/include/torch/ordered_dict.h#L345

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,135 @@
+  bool contains(const std::string& key) {
+    return parameters_.contains(key);
+  }
+
```
Same here, can we put comments said this function will throw if key is not in the dict, call contains first to avoid throw.

And I think you need another const version here
const Tensor& get(const std::string& key) const.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,135 @@
+  Tensor& get(const std::string& key) {
+    return parameters_[key];
+  }
+
```
Same here, add exception throw comments

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,135 @@
+
+  explicit ParameterDictImpl(
+      const torch::OrderedDict<std::string, torch::Tensor>& params) {
+    parameters_ = params;
```
This is fine.
The reason you get compile error is that parameters_ is not a direct member variable of ParameterDictImpl, it can not be initialized using initialize_list in constructor. What I meant is that you can check whether 'Module' class has a constructor which takes 'params' as argument, so you can call 'Module' class's constructor to initialize it.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,135 @@
+    return register_parameter(key, param, param.requires_grad());
+  }
+
+  /// Remove key from the ParameterDict, and return its value
```
Please add comments here said this function will throw if key is not in dict, call 'contains` before to avoid exceptions. Something like this.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/parameterdict.h:
```diff
@@ -0,0 +1,135 @@
+  Tensor& operator[](const std::string& key) {
+    return parameters_[key];
+  }
+
```
Same here, add exception throw comments

glaringlee reviewed (changes requested):
@yyn19951228 Sorry about delay. I added more comments, please take a look at. Thanks a lot.

arktrail committed:
ParameterDict V5

arktrail committed:
ParameterDict V5

arktrail committed:
ParameterDict V5

glaringlee reviewed (changes requested):
@yyn19951228 some nit things, then we are good to go. Thanks a lot for your help!

yyn19951228 commented:
Hi @glaringlee , I have updated these comments! Please check if there is anything left, thank you very much!

zasdfgbnm committed:
Remove cpu vec256 for std::complex (#39830)

Summary:
std::complex is gone. We are now using c10::complex
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39830

Differential Revision: D22252066

Pulled By: malfet

fbshipit-source-id: cdd5bb03ec66825d82177d609cbcf0738922dba0

wanchaol committed:
[doc] fix autograd doc subsubsection display issue (#40582)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40582

There's a misuse in the `requires_grad` with ~~~~~~, "~~~~" is not a official section marker, change it to "^^^^^" to denote subsubsections, also fix the other places where we should use subsection "-----" instead of subsubsection "^^^^"

see https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections

Before:
<img width="712" alt="rst_before" src="https://user-images.githubusercontent.com/9443650/85789835-2226fa80-b6e4-11ea-97b6-2b19fdf324a4.png">
After:
<img width="922" alt="rst_after" src="https://user-images.githubusercontent.com/9443650/85789856-281cdb80-b6e4-11ea-925f-cb3f4ebaa2bf.png">

Test Plan: Imported from OSS

Differential Revision: D22245747

Pulled By: wanchaol

fbshipit-source-id: 11548ed42f627706863bb74d4269827d1b3450d4

shoumikhin committed:
[papaya][aten] Fix compiler error: loop variable 'tensor' is always a copy because the range of type 'c10::List<at::Tensor>' does not return a reference. (#40599)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40599

.

Test Plan: CI

Reviewed By: smessmer

Differential Revision: D22246106

fbshipit-source-id: a5d0535e627b9f493fca7234dcfc15c521b0ed7f

glaringlee reviewed (changes requested):
@yyn19951228 Thanks a lot, this is fine now. 
Can you rebase this with pytorch master branch? I will approve this then.

iseeyuan committed:
[Selective Build] Apply query-based on instrumentation_tests

Summary:
1. Modularize some bzl files to break circular buck load
2. Use query-based on instrumentation_tests

(Note: this ignores all push blocking failures!)

Test Plan: CI

Reviewed By: kwanmacher

Differential Revision: D22188728

fbshipit-source-id: affbabd333c51c8b1549af6602c6bb79fabb7236

malfet committed:
Add C++ standard version check to top level headers (#40510)

Summary:
Remove `-std=c++14` flag from `utils.cmake`, since PyTorch C++ API can be invoked by any compiler compliant with C++14 standard or later
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40510

Differential Revision: D22253313

Pulled By: malfet

fbshipit-source-id: ff731525868b251c27928fc98b0724080ead9be2

ezyang committed:
Precompute entries in dispatch tables (#40512)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40512

Fixes https://github.com/pytorch/pytorch/issues/32454

The heart of this diff is changing this:

```
inline const KernelFunction& Dispatcher::dispatch_(const DispatchTable& dispatchTable, DispatchKey dispatchKey) c
nst {
  const KernelFunction* backendKernel = dispatchTable.lookup(dispatchKey);

  if (nullptr != backendKernel) {
    return *backendKernel;
  }

  const auto& backendFallbackKernel = backendFallbackKernels_[dispatchKey];
  if (backendFallbackKernel.isValid()) {
    return backendFallbackKernel;
  }

  const KernelFunction* catchallKernel = dispatchTable.lookupCatchallKernel();
  if (C10_LIKELY(nullptr != catchallKernel)) {
    return *catchallKernel;
  }

  reportError(dispatchTable, dispatchKey);
}
```

to this:

```
const KernelFunction& OperatorEntry::lookup(DispatchKey k) const {
  const auto& kernel = dispatchTable_[static_cast<uint8_t>(k)];
  if (C10_UNLIKELY(!kernel.isValid())) {
    reportError(k);
  }
  return kernel;
}
```

The difference is that instead of checking a bunch of places to find the
right kernel to use for an operator, all of the operators are
precomputed into dispatchTable_ itself (so you don't have to consult
anything else at runtime.)  OperatorEntry::computeDispatchTableEntry
contains that computation (which is exactly the same as it was before.)
By doing this, we are able to substantially simplify many runtime
components of dispatch.

The diff is fairly large, as there are also some refactors interspersed
with the substantive change:

- I deleted the DispatchTable abstraction, folding it directly into
  OperatorEntry.  It might make sense to have some sort of DispatchTable
  abstraction (if only to let you do operator[] on DispatchKey without
  having to cast it to integers first), but I killed DispatchTable to
  avoid having to design a new abstraction; the old abstraction wasn't
  appropriate for the new algorithm.

- I renamed OperatorEntry::KernelEntry to AnnotatedKernel, and use it
  to store backend fallbacks as well as regular kernel registrations
  (this improves error messages when you incorrectly register a backend
  fallback twice).

- I moved schema_ and debug_ into an AnnotatedSchema type, to make the
  invariant clearer that these are set together, or not at all.

- I moved catch-all kernels out of kernels_ into its own property
  (undoing a refactor I did before).  The main reason I did this was
  because our intended future state is to not have a single catch-all,
  but rather possibly multiple catch-alls which fill-in different
  portions of the dispatch table.  This may change some more in
  the future: if we allow registrations for multiple types of
  catch alls, we will need a NEW data type (representing bundles
  of dispatch keys) which can represent this case, or perhaps
  overload DispatchKey to also record these types.

The key changes for precomputation:

- OperatorEntry::updateDispatchTable_ is now updated to fill in the
  entry at a DispatchKey, considering both kernels (what it did
  before) as well as catch-all and backend fallback.  There is also
  OperatorEntry::updateDispatchTableFull_ which will update the
  entire dispatch table (which is necessary when someone sets a
  catch-all kernel).  OperatorEntry::computeDispatchTableEntry
  holds the canonical algorithm specifying how we decide what
  function will handle a dispatch key for the operator.

- Because dispatch table entry computation requires knowledge of
  what backend fallbacks are (which is recorded in Dispatcher,
  not OperatorEntry), several functions on OperatorEntry now
  take Dispatcher as an argument so they can query this information.

- I modified the manual boxing wrapper invariant: previously, kernels
  stored in kernels_ did NOT have manual boxing wrappers and this
  was maintained by DispatchTable.  Now, we just ALWAYS maintain
  manual boxing wrappers for all KernelFunctions we store.

- DispatchKeyExtractor is greatly simplified: we only need to maintain
  a single per-operator bitmask of what entries are fallthrough
  (we don't need the global bitmask anymore).

- Introduced a new debugging 'dumpComputedTable' method, which prints
  out the computed dispatch table, and how we computed it to be some way.
  This was helpful for debugging cases when the dispatch table and
  the canonical metadata were not in sync.

Things that I didn't do but would be worth doing at some point:

- I really wanted to get rid of the C10_UNLIKELY branch for
  whether or not the KernelFunction is valid, but it looks like
  I cannot easily do this while maintaining good error messages.
  In principle, I could always populate a KernelFunction which
  errors, but the KernelFunction needs to know what the dispatch
  key that is missing is (this is not passed in from the
  calling convention).  Actually, it might be possible to do
  something with functors, but I didn't do it here.

- If we are going to get serious about catchalls for subsets of
  operators, we will need to design a new API for them.  This diff
  is agnostic to this question; we don't change public API at all.

- Precomputation opens up the possibility of subsuming DispatchStub
  by querying CPU capability when filling in the dispatch table.
  This is not implemented yet. (There is also a mild blocker here,
  which is that DispatchStub is also used to share TensorIterator
  configuration, and this cannot be directly supported by the
  regular Dispatcher.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D22236352

Pulled By: ezyang

fbshipit-source-id: d6d90f267078451816b1899afc3f79737b4e128c

ezyang committed:
Generalize Python dispatcher testing API; disallow overwriting fallback (#40469)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40469

- The old testing interface C._dispatch_import was based off the old
  c10::import variation, which meant the API lined up in a strange
  way with the actual torch/library.h.  This diff reduces the
  differences by letting you program the Library constructor directly.

- Using this newfound flexibility, we add a test for backend fallbacks
  from Python; specifically testing that we disallow registering a
  backend fallback twice.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Differential Revision: D22236351

Pulled By: ezyang

fbshipit-source-id: f8365e3033e9410c7e6eaf9f78aa32e1f7d55833

arktrail committed:
rebase

yyn19951228 commented:
Hi @glaringlee, could you please check if my rebase breaks something? I am afraid that some of my operators might break some comments history. 
If everything goes fine, I am happy to ship this diff and work on the `ParameterList`.

Commit:
update schema to reflect aliasing behavior (#39794)

Summary:
Fixes: https://github.com/pytorch/pytorch/issues/38555

I did an audit of `native_functions.yaml` and found several functions in addition to `reshape` which were not reporting that they could alias:

```
torch.jit.script
def foo(t: torch.Tensor):
    new_value = torch.tensor(1, dtype=t.dtype, device=t.device)

    t.flatten()[0] = new_value
    t.reshape(-1)[1] = new_value
    t.view_as(t)[2] = new_value
    t.expand_as(t)[3] = new_value
    t.reshape_as(t)[4] = new_value
    t.contiguous()[5] = new_value
    t.detach()[6] = new_value

    return t
```

Currently none of the values are assigned after dead code elimination, after this PR all are. (And the JIT output matches that of eager.)

I don't think this needs to be unit tested; presumably the generic machinery already is and this just brings these ops under the same umbrella.

**BC-breaking note**: This updates the native operator schema and the aliasing rules for autograd. JIT passes will no longer incorrectly optimize mutations on graphs containing these ops, and inplace ops on the result of `flatten` will now properly be tracked in Autograd and the proper backward graph will be created.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39794

Differential Revision: D22008358

Pulled By: robieta

fbshipit-source-id: 9d3ff536e58543211e08254a75c6110f2a3b4992

jeffdaily committed:
skip_if_rocm test_rnn in test_c10d_spawn.py (#40577)

Summary:
Test was added a few months back in https://github.com/pytorch/pytorch/issues/36503 but recently became flaky for ROCm.

CC ezyang xw285cornell sunway513
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40577

Differential Revision: D22258196

Pulled By: ezyang

fbshipit-source-id: 8a22b0c17b536b3d42d0382f7737df0f8823ba08

zasdfgbnm committed:
build docker image for CUDA11 (#40534)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/40534

Differential Revision: D22258874

Pulled By: seemethere

fbshipit-source-id: 1954a22ed52e1a65caf89725ab1db9f40ff917b8

malfet committed:
Further reduce windows build/test matrix (#40592)

Summary:
Switch windows CPU testers from `windows.xlarge` to `windows.medium` class.
Remove VS 14.16 CUDA build
Only do smoke force-on-cpu tests using VS2019+CUDA10.1 config.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40592

Differential Revision: D22259351

Pulled By: malfet

fbshipit-source-id: f934ff774dfc7d47f12c3da836ca314c12d92208

mitchellspryn committed:
Fix illegal opcode bug in caffe2 (#40584)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40584

Also patch [this github issue](https://github.com/pytorch/pytorch/issues/33124)
involving an illegal assembly instruction in 8x8-dq-aarch64-neon.S.

Test Plan:
Build binaries, copy to shaker, run executables. Also run all
existing caffe tests.

Reviewed By: kimishpatel

Differential Revision: D22240670

fbshipit-source-id: 51960266ce58699fe6830bcf75632b92a122f638

Krovatkin committed:
properly skip legacy tests regardless of the default executor (#40381)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/40381

Differential Revision: D22173938

Pulled By: Krovatkin

fbshipit-source-id: 305fc4484977e828cc4cee6e053a1e1ab9f0d6c7

lw committed:
Update TensorPipe submodule (#40614)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40614

This update pulls in a oneliner fix, which sets the TCP_NODELAY option on the TCP sockets of the UV transport. This leads to exceptional performance gains in terms of latency, with about a 25x improvement in one simple benchmark. This thus resolves a regression that TensorPipe had compared to the ProcessGroup agent and, in fact, ends up beating it by 2x.

The benchmark I ran is this, with the two endpoints pinned to different cores of the same machine:
```
torch.jit.script
def remote_fn(t: int):
    return t

torch.jit.script
def local_fn():
    for _ in range(1_000_000):
        fut = rpc.rpc_async("rhs", remote_fn, (42,))
        fut.wait()
```

And the average round-trip time (one iteration) is:
- TensorPipe with SHM: 97.2 us
- TensorPipe with UV _after the fix_: 205us
- Gloo: 440us
- TensorPipe with UV _before the fix_: 5ms

Test Plan: Ran PyTorch RPC test suite

Differential Revision: D22255393

fbshipit-source-id: 3f6825d03317d10313704c05a9280b3043920507

glaringlee commented:
@yyn19951228 
Hi, I think there are some problems with your rebase. 
This PR should only have your changes, but it is now have 28 files changes :).
If you have problems on rebasing this correctly, you can open a new pr with the latest pytorch and put your code in.
I will close this one and approve your new PR.

Commit:
[JIT] Remove dead store in exit_transforms.cpp (#40611)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40611

This commit removes a dead store in `transformWith` of exit_transforms.cpp.

Test Plan: Continuous integration.

Reviewed By: suo

Differential Revision: D22254136

fbshipit-source-id: f68c4625f7be8ae29b3500303211b2299ce5d6f6

smessmer committed:
Back out "Revert D21581908: Move TensorOptions ops to c10" (#40595)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40595

ghstack-source-id: 106691774

Test Plan: waitforsandcastle

Differential Revision: D22247729

fbshipit-source-id: 14745588cae267c1e0cc51cd9541a9b8abb830e5

malfet committed:
Add special rules to launch docker image with RocM (#40632)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/40632

Differential Revision: D22262316

Pulled By: malfet

fbshipit-source-id: 3d525767bfbfc8e2497541849d85cabf0379a43b

dreiss committed:
Add a utility function for bundling large input tensors (#37055)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37055

Sometimes it's okay to bundle a large example input tensor with a model.
Add a utility function to make it easy for users to do that *on purpose*.

Test Plan: Unit test.

Differential Revision: D22264239

Pulled By: dreiss

fbshipit-source-id: 05c6422be1aa926cca850f994ff1ae83c0399119

dreiss committed:
Add channels-last support to bundled_inputs (#36764)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/36764

This allows bundling inputs that are large uniform buffers in
channels-last memory format.

Test Plan: Unit test.

Differential Revision: D21142660

Pulled By: dreiss

fbshipit-source-id: 31bbea6586d07c1fd0bcad4cb36ed2b8bb88a7e4

malfet committed:
Ignore error code from `apt-get purge` (#40631)

Summary:
This replicates the pattern of other "do for luck" commands.
Prep change to add RocM to CircleCI.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40631

Differential Revision: D22261707

Pulled By: malfet

fbshipit-source-id: 3dadfa434deab866a8800715f3197e84169cf43e

xwang233 committed:
[Reland][doc] Add overflow notice for cuFFT on half precision (#40551)

Summary:
Reland of https://github.com/pytorch/pytorch/issues/35594
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40551

Reviewed By: ezyang

Differential Revision: D22249831

Pulled By: ngimel

fbshipit-source-id: b221b3c0a490ccaaabba50aa698a2490536e0917

changjishi committed:
Port /test/cpp_extensions/rng_extension.cpp to new operator registration API (#39459)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39459

Update to this PR: this code isn't going to fully solve https://github.com/pytorch/pytorch/issues/37010. The changes required for 37010 is more than this PR initially planned. Instead, this PR switches op registration of rng related tests to use the new API (similar to what was done in #36925)

Test Plan:
1) unit tests

Imported from OSS

Reviewed By: ezyang

Differential Revision: D22264889

fbshipit-source-id: 82488ac6e3b762a756818434e22c2a0f9cb9dd47

Commit:
FP16 rounding-to-nearest for row-wise SparseAdagrad fusion (#40466)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40466

Extend row wise sparse Adagrad fusion op to FP16 (rounding-to-nearest) for PyTorch.

Reviewed By: jianyuh

Differential Revision: D22003571

fbshipit-source-id: e97e01745679a9f6e7b0f81ce5a6ebf4d4a1df41

Commit:
[1/n] Allow dense NaN value in dper raw input processor output

Summary:
## TLDR
Support using NaN default value for missing dense features in RawInputProcessor for *DPER2*. In preparation for subsequent support for null flag features in *compute meta*. For train_eval this is already supported in DPER3 and we do not plan to support this in DPER2 train eval.
## Overview
Intern project plan to support adding dense flags for missing feature values instead of replacing with zero.

Project plan :
https://docs.google.com/document/d/1OsPUTjpJycwxWLCue3Tnb1mx0uDC_2KKWvC1Rwpo2NI/edit?usp=sharing

## Code paths:
See https://fb.quip.com/eFXUA0tbDmNw for the call stack for all affected code paths.

Test Plan:
# A. DPER3 blob value inspection
## 1. Build local bento kernel in fbcode folder
`buck build mode/dev-nosan //bento/kernels:bento_kernel_ads_ranking`

## 2. Use kernel `ads_ranking (local)` to print dense feature blob values
n280239

## 2.1 Try `default_dense_value = "0.0"` (default)
```
preproc_6/feature_preproc_6/dper_feature_processor_7/raw_input_proc_7/float_feature_sparse_to_dense_7/float_features [[0.       ]
 [0.       ]
 [0.       ]
 [0.       ]
 [0.       ]
 [0.       ]
 [0.       ]
 [1.       ]
 [1.7857143]
 [1.7777778]
 [1.       ]
 [0.       ]
 [0.5625   ]
 [0.       ]
 [0.       ]
 [0.8      ]
 [0.       ]
 [1.       ]
 [0.56     ]
 [0.       ]]
```
## 2.2 Try `default_dense_value = "123"`
```
preproc_2/feature_preproc_2/dper_feature_processor_3/raw_input_proc_3/float_feature_sparse_to_dense_3/float_features [[123.       ]
 [123.       ]
 [123.       ]
 [123.       ]
 [123.       ]
 [123.       ]
 [123.       ]
 [  1.       ]
 [  1.7857143]
 [  1.7777778]
 [  1.       ]
 [123.       ]
 [  0.5625   ]
 [123.       ]
 [123.       ]
 [  0.8      ]
 [123.       ]
 [  1.       ]
 [  0.56     ]
 [123.       ]]
```
## 2.3 Try `default_dense_value = float("nan")`
```
RuntimeError: [enforce fail at enforce_finite_op.h:40] std::isfinite(input_data[i]). Index 0 is not finite (e.g., NaN, Inf): -nan (Error from operator:
input: "unary_4/logistic_regression_loss_4/average_loss_4/average_loss" name: "" type: "EnforceFinite" device_option { random_seed: 54 })
```
which is expected due to nan input.

# B. Unit test
`buck test  fblearner/flow/projects/dper/tests/preprocs:raw_feature_extractor_test`

https://www.internalfb.com/intern/testinfra/testconsole/testrun/5348024586274923/

{F241336814}

Differential Revision: D21961595

fbshipit-source-id: 3dcb153b3c7f42f391584f5e7f52f3d9c76de31f

arktrail committed:
fix rebase

yyn19951228 commented:
Hi @glaringlee , sorry about that! I have created a new PR #40654, and could you please check that? Thank you very much!

glaringlee commented:
Close this PR since it is hard to rebase in this PR.
@yyn19951228 opened a new PR #40654 with same change.
